### ISSUE TITLE
Spec doesn't mention handling of leap seconds

### ISSUE THREAD

**Issue by heinrich5991** [matrix-org/matrix-spec#1626]
From [the Matrix specification](https://spec.matrix.org/v1.7/#timestamps):
> ### Timestamps
>
> Unless otherwise stated, timestamps are measured as milliseconds since the Unix epoch. Throughout the specification this may be referred to as POSIX, Unix, or just “time in milliseconds”.

**Issue**
This does not talk about leap seconds, or rather has two conflicting statements about them. The number of milliseconds since 1970-01-01 00:00:00 includes leap seconds, but [Unix time](https://en.wikipedia.org/w/index.php?title=Unix_time&oldid=1168330527) usually does not. In Unix time, each day is fixed to have 86400 seconds.

How should the matrix spec be interpreted wrt. that? Can timestamps repeat (Unix time) or can they not (most implementations would probably be wrong since that's not Unix time anymore).

**Comment by KitsuneRal** [comment]
Timestamps can repeat regardless of leap seconds, simply because several events can be sent within the same second. From the practical perspective - do you know any library or computer standard that represents a point in time as a single integer number _since Unix epoch_ that is not Unix time? I don't think there is one.

For the record, the current text in turn is driven by the fact that originally "Unix time" is defined in seconds, not milliseconds. Just replacing it with "Unix time" would lose that important point.

**Comment by richvdh** [comment]
As @KitsuneRal says, timestamps in Matrix ignore leap seconds. The same, in fact, as [`Date.now()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date) in Javascript, and in many other languages.

> Timestamps can repeat regardless of leap seconds, simply because several events can be sent within the same second. 

That's not quite the same thing though. The question is: "are you allowed to emit an event with a timestamp of 1483228799000 after you've emitted on with a timestamp of 1483228799999, because 2016-12-31 23:59:59 happens twice?"  And I think the answer is: "shrug". I don't see any real reason why not, or anything that will stop working if you do so. (Indeed, I'm not exactly certain what Synapse will do; and if you set Synapse up in worker mode, with workers on different servers with different clocks, then it *will* do that sort of thing.)

In short:
 * Matrix timestamps ignore leap seconds.
 * Dealing with leap seconds is an implementation decision.

A PR clarifying that in the spec would be appreciated.
